src/include/OpenImageIO/fmath.h

```diff
@@ -1742,4 +1753,181 @@ interpolate_linear (float x, array_view_strided<const float> y)
+//
+// For now, the selection of 'fast' instrinsics is controlled via compiler
+// flags.
+
+#ifdef __CUDACC__
+__device__ inline float fast_sin (float x) {
+    return __sinf(x);
+}
+
+__device__ inline float fast_cos (float x) {
+    return __cosf(x);
+}
+
+__device__ inline void fast_sincos (float x, float* sine, float* cosine) {
+    __sincosf(x, sine, cosine);
+}
+
+__device__ inline float fast_tan (float x) {
+    return __tanf(x);
+}
+
+__device__ inline float fast_sinpi (float x) {
+    return __sinf(float(M_PI)) * x;
+}
+
+__device__ inline float fast_cospi (float x) {
+    return __cosf(float(M_PI)) * x;
+}
+
+__device__ inline float fast_acos (float x) {
+    return acosf(x);
+}
+
+__device__ inline float fast_asin (float x) {
+    return asinf(x);
+}
+
+__device__ inline float fast_atan (float x) {
+    return atanf(x);
+}
+
+__device__ inline float fast_atan2 (float y, float x) {
+    return atan2f( y, x );
+}
+
+__device__ inline float fast_log2 (float xval) {
+    return __log2f( xval );
+}
+
+__device__ inline float fast_log (float x) {
+    return __logf(x);
+}
+
+__device__ inline float fast_log10 (float x) {
+    return __log10f(x);
+}
+
+__device__ inline float fast_logb (float x) {
+    return logbf(x);
+}
+
+__device__ inline float fast_log1p (float x) {
+    return log1pf(x);
+}
+
+__device__ inline float fast_exp (float x) {
+    return __expf(x);
+}
+
+__device__ inline float fast_exp2 (float x) {
+    return exp2f(x);
+}
+
+__device__ inline float fast_exp10 (float x) {
+    return __exp10f(x);
+}
+
+__device__ inline float fast_expm1 (float x) {
+    return expm1f(x);
+}
+
+__device__ inline float fast_erf (float x) {
+    return erff(x);
+}
+
+__device__ inline float fast_erfc (float x) {
+    return erfcf(x);
+}
+
+__device__ inline float fast_sinh (float x) {
+    return sinhf(x);
+}
+
+__device__ inline float fast_cosh (float x) {
+    return coshf(x);
+}
+
+__device__ inline float fast_tanh (float x) {
+    return tanhf(x);
+}
+
+__device__ inline float fast_sqrt (float x) {
+    return __fsqrt_rn(x);
+}
+
+__device__ inline float fast_inversesqrt (float x) {
+    return __frsqrt_rn(x);
+}
+
+__device__ inline float safe_sqrt (float x) {
+    return x >= 0.0f ? __fsqrt_rn(x) : 0.0f;
+}
+
+__device__ inline float safe_inversesqrt (float x) {
+    return x > 0.0f ? __frsqrt_rn(x) : 0.0f;
+}
+
+// (end CUDA intrinsics)
+////////////////////////////////////////////////////////////////////////////
+
+////////////////////////////////////////////////////////////////////////////
+// 'Safe' device functions & helpers
+
+template <typename IN_TYPE, typename OUT_TYPE> __device__
+inline OUT_TYPE device_bit_cast (const IN_TYPE in) {
+    // NOTE: this is the only standards compliant way of doing this type of casting,
+    // luckily the compilers we care about know how to optimize away this idiom.
+    OUT_TYPE out;
+    memcpy (&out, &in, sizeof(IN_TYPE));
+    return out;
+}
+
+__device__ inline float safe_acos (float x) {
```

Comment:
Again, here, if this code for safe_acos is identical to the other one, I think that instead of two definitions (one with `__device__` and one without), we should have a single definition and a macro for DEVICE that does the right thing based on the compiler.